packages/mongodb-access-adapter/datagrip-access-adapter/src/main/kotlin/com/mongodb/jbplugin/accessadapter/datagrip/adapter/DataGripMongoDbDriver.kt

```diff
@@ -33,60 +37,96 @@ import kotlinx.coroutines.withTimeout
  */
 internal class DataGripMongoDbDriver(
     private val project: Project,
-    private val dataSource: LocalDataSource
+    private val dataSource: LocalDataSource,
 ) : MongoDbDriver {
     private val gson = Gson()
+    private val jsonWriterSettings =
+        JsonWriterSettings
+            .builder()
+            .outputMode(JsonMode.EXTENDED)
+            .indent(false)
+            .build()
+
+    private fun String.encodeForJs(): String = Encode.forJavaScript(this)
+
+    private fun Bson.toJson(): String = this.toBsonDocument().toJson(jsonWriterSettings)
+
+    override suspend fun connectionString(): ConnectionString = ConnectionString(dataSource.url!!)
 
     override suspend fun <T : Any> runCommand(
-        command: Document,
+        database: String,
+        command: Bson,
         result: KClass<T>,
-        timeout: Duration
-    ): T = withContext(
-        Dispatchers.IO
-    ) {
-        runQuery(
-            """db.runCommand(${command.toJson()})""",
-            result,
-            timeout
-        )[0]
-    }
+        timeout: Duration,
+    ): T =
+        withContext(
+            Dispatchers.IO,
+        ) {
+            runQuery(
+                """
+                db.getSiblingDB("${database.encodeForJs()}")
+                  .runCommand(EJSON.parse(`${command.toJson()}`))
+                """.trimIndent(),
+                result,
+                timeout,
+            )[0]
+        }
 
     override suspend fun <T : Any> findOne(
         namespace: Namespace,
-        query: Document,
-        options: Document,
+        query: Bson,
+        options: Bson,
         result: KClass<T>,
-        timeout: Duration
-    ): T? = withContext(Dispatchers.IO) {
-        runQuery(
-            """db.getSiblingDB("${namespace.database}")
-                 .getCollection("${namespace.collection}")
-                 .findOne(${query.toJson()}, ${options.toJson()}) """.trimMargin(),
-            result,
-            timeout
-        ).getOrNull(0)
-    }
+        timeout: Duration,
+    ): T? =
+        withContext(Dispatchers.IO) {
+            runQuery(
+                """db.getSiblingDB("${namespace.database.encodeForJs()}")
+                 .getCollection("${namespace.collection.encodeForJs()}")
+                 .findOne(EJSON.parse(`${query.toJson()}`), EJSON.parse(`${options.toJson()}`)) 
```

Comment:
Does this work if the query contains a `` ` ``? Should this be 
```suggestion
                 .findOne(EJSON.parse("${query.toJson().encodeForJs()}"), EJSON.parse("${options.toJson()}.encodeForJs()")) 
```

Reply:
Hmmm toJson already encodes the BSON document using the JsonWriterSettings, so that part should be safe. About using double quotes or back quotes, I think both should be fine? But I can change it to double quotes.

Reply:
I don't care about the specific quotes, just making sure that we can't end up with e.g. `` EJSON.parse(`{ "foo": "bar`baz" }`) `` here (because JSON writers in general don't need to escape backticks in strings)

Reply:
Makes sense, I'll change it to double quotes to make sure we don't get some weird scenario where someone can inject a query here.

Thanks!

Reply:
Ok, but now you definitely have issues because the JSON text will invariably contain double quotes (hence the suggestion to escape those, so we get reliably and "obviously correct" injections)

Reply:
Yeah, you are right, I didn't realise. I'm going to force encoding the JSON that we convert using the OWASP encoder and use double quotes. This way it should be entirely safe.